### PR TITLE
SSE-3231 Frontend for User attributes

### PR DESCRIPTION
**Pull request by MartinH-DI**
# Onboarding Feature Deployment

> [!WARNING]
> Pull requests merged to `main` will be released to production, please ensure the checklist below is complete

Before any work can be merged to main in must meet the definition of done and be ready to deploy. While many of these tasks will be automated, the reviewers must take the responsibility of confirming the checklist below has been completed before this ticket can be merged.

## Checklist

-   [ ] this pull request meets the acceptance criteria of the ticket

-   [ ] this branch is up-to-date with the main branch

    `git fetch --all && git rebase origin/main`

-   [ ] tests have been written to cover any new or updated functionality

-   [ ] new configuration parameters have been deployed to all environments, see [configuration management](https://govukverify.atlassian.net/l/cp/N7q3Vh3r).

-   [ ] all external infrastructure dependencies have been updated in all environments

## Changes

[ _please list the changes this pull request is making_ ]

### `Added` for new features

New User Attributes section added to the Service management page

### `Changed` for changes in existing functionality

User Attributes section is moved to new section and renamed Scopes, with associated changes to the detail screen and tests.

### CODE REVIEW

File: express/src/views/layouts/base.njk [hunk 1]
```diff
@@ -13,6 +13,8 @@
 
 {% set govUkGridClass = gridClass or "govuk-grid-column-two-thirds" %}
 
+{% set govUkGridClass = gridClass or "govuk-grid-column-two-thirds" %}
```
Comment (BorisP-GDS) [inline express/src/views/layouts/base.njk:16]: This line is not needed as it duplicates line 14

**Review by BorisP-GDS — COMMENTED** [review]
Changes required as per my comments above.

File: express/src/controllers/clients.ts [hunk 2]
```diff
@@ -33,29 +36,29 @@ export const showClient: RequestHandler = async (req, res) => {
...
+        scopesRequired: client.scopes,
         ...(client.token_endpoint_auth_method === "client_secret_post"
             ? {client_secret: client.client_secret ?? ""}
             : {userPublicKey: userPublicKey}),
         backChannelLogoutUri: client.back_channel_logout_uri,
         sectorIdentifierUri: userSectorIdentifierUri,
         postLogoutRedirectUris: client.postLogoutUris,
-        claims: client.identity_verification_enabled && client.hasOwnProperty("claims") ? client.claims : [],
-        id_token_signing_algorithm: client.hasOwnProperty("id_token_signing_algorithm") ? client.id_token_signing_algorithm : "",
+        postLogoutRedirectUrls: client.postLogoutUris.join(" "),
```
[the remark this answers is not in the document]
Comment (MartinH-DI) [reply express/src/controllers/clients.ts:46]: I think this was a mistake on merging: I'll remove it.

File: express/src/controllers/clients.ts [hunk 3]
```diff
@@ -787,7 +853,7 @@ export const processEnterContactEmailForm: RequestHandler = async (req, res) =>
     if (result.isValid) {
         const updateContactsResult = addContact(contactToAdd, contacts);
 
-        if (typeof updateContactsResult === "string" && updateContactsResult === "This contact has already been added") {
+        if (typeof updateContactsResult === "string" && updateContactsResult === "Contact already exists") {
```
[the remark this answers is not in the document]
Comment (MartinH-DI) [reply express/src/controllers/clients.ts:856]: I think this was a mistake on merging: I'll revert it.

File: express/src/views/layouts/base.njk (hunk 1 above)
Comment (MartinH-DI) [reply express/src/views/layouts/base.njk:16]: I think this was a mistake on merging: I'll remove it.

File: express/src/controllers/clients.ts [hunk 4]
```diff
@@ -292,6 +306,58 @@ export const processChangePublicKeyForm: RequestHandler = async (req, res) => {
     res.redirect(`/services/${req.context.serviceId}/clients`);
 };
 
+export const showEnterClientSecretHashForm: RequestHandler = (req, res) => {
```
Comment (rafal-surowiec-co) [inline express/src/controllers/clients.ts:309]: Martin, can you add unit tests ?

Comment (MartinH-DI) [reply express/src/controllers/clients.ts:308]: Unit tests have now been added.

File: express/src/controllers/clients.ts [hunk 5]
```diff
@@ -451,23 +457,24 @@ export const processRemoveRedirectUriFrom: RequestHandler = async (req, res) =>
...
+    res.render("clients/change-scopes.njk", {
+        selectedScopes: req.query.scopes?.toString().split(" "),
         serviceId: req.context.serviceId,
         selfServiceClientId: req.params.selfServiceClientId,
         clientId: req.params.clientId
     });
 };
 
-export const processChangeUserAttributesForm: RequestHandler = async (req, res) => {
+// eslint-disable-function @typescript-eslint/no-misused-promises
```
Comment (BorisP-GDS) [inline express/src/controllers/clients.ts:469]: We can avoid disabling the lint here. I suggest to try to remove the eslint disable line and try  : 

```
export const processChangeScopesForm: RequestHandler = async (req: Request, res: Response): Promise<void> => {
    try {
        const s4: SelfServiceServicesService = req.app.get("backing-service");
        const attributes = ["openid"];
        if (Array.isArray(req.body.scopes)) {
            attributes.push(...req.body.scopes);
        } else if (typeof req.body.scopes === "string") {
            attributes.push(req.body.scopes);
        }

        await s4.updateClient(
            nonNull(req.context.serviceId),
            req.params.selfServiceClientId,
            req.params.clientId,
            { scopes: attributes },
            nonNull(req.session.authenticationResult?.AccessToken)
        );

        req.session.updatedField = "scopes";
        res.redirect(`/services/${req.context.serviceId}/clients`);
    } catch (error) {
        console.error("Error processing change scopes form:", error);
        res.status(500).send("Internal Server Error");
    }
};
```

Comment (MartinH-DI) [reply express/src/controllers/clients.ts]: Thanks I'll give that a go. This disable doesn't work anyway: the web sonar ignores it.

Comment (MartinH-DI) [reply express/src/controllers/clients.ts]: I'm afraid that suggestion doesn't work either. It still throws it out.

**Pull request merged**